CDC: Throttle the maximum number of requests per second.

batiati opened the pull request:
Implements a `--requests-per-second-limit` option to throttle CDC `get_change_events` requests to TigerBeetle.  
Its usage is orthogonal to `--idle-interval-ms` and `--event-count-max`, allowing fine-tuning for low latency without overflowing the AMQP target queue.

This PR also includes unrelated but necessary refactors (best reviewed by commit):
- Adds `vsr.time.Timer`, an alternative to `std.time.Timer` that uses the VSR abstracted `Time` interface.  
- Replaces some usages of `std.time.Timer` with the new implementation.  
- Refactors `init` functions to avoid including dependencies in the `options` struct, enforcing recent TigerStyle updates (https://github.com/tigerbeetle/tigerbeetle/pull/3101):
>   Because dependencies like an allocator or a tracer are singletons with unique types, they should
  be threaded through constructors positionally, from the most general to the most specific.

batiati committed:
Add `vsr.time.Timer`.
An alternative to `std.time.Timer` using the abstracted `Time` interface.

batiati committed:
Use `vsr.time.Timer` instead of `std.time.Timer`.
- Propagate the dependency on the `vsr` `Time` abstraction.
- Refactor the `options: struct { ... }` pattern to not include singleton dependencies, passing them as positional parameters instead.

batiati committed:
CDC: RateLimit
RateLimit to throttle the number of requests made to TigerBeetle, and consequently the number of events sent to the AMQP target.

batiati committed:
CDC: Update the docs to add the rate limit CLI option.

batiati committed:
Fix aof.

sentientwaffle commented on src/time.zig:
```diff
@@ -184,3 +184,68 @@ test "Time monotonic smoke" {
+    /// guaranteed to be monotonic with respect to the timer's starting point.
+    fn sample(self: *Timer) Instant {
+        const current = self.time.monotonic_instant();
+        if (current.ns > self.previous.ns) {
```
I'm confused; isn't this condition guaranteed since we are using `monotonic_instant`?

batiati replied:
Nice catch! I simply vendored the `std.time.Timer` code here without noticing that the `Time` interface already provides the monotonic guard.

sentientwaffle commented on src/cdc/runner.zig:
```diff
@@ -221,14 +221,17 @@ pub const Runner = struct {
+        self.vsr_client = try Client.init(
+            allocator,
+            &self.message_pool,
+            time,
```
Nit, but I would probably swap the order of these arguments, so that we go from "most general -> most specific".

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -1065,7 +1065,14 @@ pub fn ReplicaType(
+            allocator: Allocator,
+            storage: *Storage,
+            message_pool: *MessagePool,
+            time: Time,
```
Likewise, I would order from time -> storage -> message_pool. (Most to least general.)

sentientwaffle commented on src/tigerbeetle/cli.zig:
```diff
@@ -1169,6 +1171,16 @@ fn parse_args_amqp(amqp: CLIArgs.AMQP) Command.AMQP {
+        if (requests_per_second_limit == 0) {
+            vsr.fatal(
+                .cli,
+                "--request-per-second-limit must not be zero.",
```
s/request/requests

sentientwaffle commented on src/cdc/runner.zig:
```diff
@@ -868,6 +917,69 @@ pub const Runner = struct {
+            return .ok;
+        }
+        assert(self.count > 0);
+        assert(self.count <= self.options.limit);
```
This assert could move up before the `if` statement.

sentientwaffle commented on src/cdc/runner.zig:
```diff
@@ -204,6 +212,12 @@ pub const Runner = struct {
+            // The rate limit is expressed in "requests per second":
+            RateLimit.init(time, .{ .limit = limit, .period = .seconds(1) })
+        else
+            null;
```
Rather than increasing dimensionality with "if rate limit enabled else" branches, what do you think of using a default rate limit of `maxInt requests/second`, so that at least some of the rate limit code is always exercised?

batiati committed:
review
- RateLimit with limit=maxInt
- Fix typo

batiati committed:
review: Fix `vsr.time.Timer` to take advantage of `monotonic_instant()`.
It was a vendored version of `std.time.Time` that had to track the previous instant to guarantee monotonicity.

batiati committed:
review: Reorder from most to least general: allocator, io, time, storage, message_pool.

batiati committed:
tidy: line break.
